abr_control/controllers/signals/avoid_joint_limits.py

```diff
@@ -13,22 +13,51 @@ class AvoidJointLimits(Signal):
     Parameters
     ----------
     min_joint_angles : np.array
-      the lower bound on joint angles [radians]
+        the lower bound on joint angles [radians]
     max_joint_angles : np.array
-      the upper bound on joint angles [radians]
+        the upper bound on joint angles [radians]
     max_torque : np.array, optional (Default: 1)
-      the maximum force to push away with [Nm]
+        the maximum force to push away with [Nm]
+    cross_zero: List of Boolean, Optional
+        (Default:[False, False, False, False, False, False])
+        select whether the section of 2pi you want to work in crosses
+        the zero-2pi boundary. Set to true if it does to account for odd
+        behaviour when we jump from 0 to 2pi and vice versa
+    gradient: List of Boolean, Optional
+        (Default:[False, False, False, False, False, False])
+        set if you want a gradual push from the limit, or a wall once reached
+        True for gradient of force, False for a sudden opposing force (Wall)
 
     NOTE: use None as a placeholder for joints that have no limits
     """
 
     def __init__(self, robot_config,
                  min_joint_angles, max_joint_angles,
-                 max_torque=None):
+                 max_torque=None,
+                 cross_zero=[False, False, False, False, False, False],
+                 gradient=[False, False, False, False, False, False]):
+        # shift limits to -pi to pi range
+        for ii in range(0,len(min_joint_angles)):
+            if min_joint_angles[ii] is not None:
+                min_joint_angles[ii] = min_joint_angles[ii] - np.pi
+            if max_joint_angles[ii] is not None:
+                max_joint_angles[ii] = max_joint_angles[ii] - np.pi
+
+        self.cross_zero = np.array(cross_zero)
+        self.gradient = np.array(gradient)
 
         self.min_joint_angles = np.asarray(min_joint_angles, dtype='float32')
         self.max_joint_angles = np.asarray(max_joint_angles, dtype='float32')
 
+        # flip in this case so math matches normal case
+        temp_min = np.copy(self.min_joint_angles)
+        temp_max = np.copy(self.max_joint_angles)
+        self.max_joint_angles[cross_zero] = temp_min[cross_zero]
+        self.min_joint_angles[cross_zero] = temp_max[cross_zero]
+
+        print('===MIN LIMITS===: ', min_joint_angles)
```

Comment:
should print to the warning log, i think

Reply:
Sounds good to me, do we have a warning log setup for control? The one I made is just for the jaco interface.

Reply:
or actually, can we just remove it?